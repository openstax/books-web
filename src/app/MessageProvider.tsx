import { shouldPolyfill } from '@formatjs/intl-pluralrules/should-polyfill';
import React from 'react';
import { IntlProvider } from 'react-intl';
import enMessages from './messages/en';

// https://formatjs.io/docs/polyfills/intl-pluralrules/#dynamic-import--capability-detection
async function polyfill(locale: 'en') {
  if (shouldPolyfill()) {
    await import('@formatjs/intl-pluralrules/polyfill');
  }

  // boolean added by the polyfill
  if ((Intl.PluralRules as (typeof Intl.PluralRules & {polyfilled?: boolean})).polyfilled) {
    switch (locale) {
      case 'en':
        await import('@formatjs/intl-pluralrules/locale-data/en');
        break;
      default:
        throw new Error('unsupported locale in messageProvider');
    }
  }
}

polyfill('en');

// tslint:disable-next-line:variable-name
const MessageProvider: React.SFC<{onError?: () => void}> = (props) =>
  <IntlProvider onError={props.onError} locale='en' messages={enMessages}>{props.children}</IntlProvider>;

export default MessageProvider;
