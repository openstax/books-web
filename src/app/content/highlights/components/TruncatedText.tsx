import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { linkStyle } from '../../../components/Typography';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { cardPadding } from '../constants';

interface Props {
  text: string;
  isFocused: boolean;
  className?: string;
}

// tslint:disable-next-line:variable-name
const Link = styled.span`
  ${textStyle}
  ${linkStyle}
  text-align: left;
  cursor: pointer;
  border: none;
  padding: 0;
  margin: ${cardPadding / 2}rem 0 0 0;
  background: none;
  font-size: 1.2rem;
  line-height: 1.6rem;
`;

// tslint:disable-next-line:variable-name
const NoteText = ({text, isFocused, className}: Props) => {
  const noteTextRef = React.useRef<HTMLElement>(null);
  const [showLink, setShowLink] = React.useState<boolean>(false);

  React.useEffect(() => {
    if (noteTextRef.current && noteTextRef.current.scrollHeight > noteTextRef.current.offsetHeight) {
      setShowLink(true);
    } else {
      setShowLink(false);
    }
  }, [isFocused]);

  return <React.Fragment>
    <p ref={noteTextRef} className={className}>{text}</p>
    {showLink && <FormattedMessage id='show more'>
      {(msg: Element | string) => <Link>{msg}</Link>}
    </FormattedMessage>}
  </React.Fragment>;
};

const lineHeight = 1.8;
export default styled(NoteText)`
  ${textStyle}
  font-size: 1.4rem;
  line-height: ${lineHeight}rem;
  margin: 0;
  padding: 0;

  ${(props: Props) => props.isFocused && css`
    + ${Link} {
      display: none;
    }
  `}
  ${(props: Props) => !props.isFocused && css`
    text-overflow: ellipsis;
    overflow: hidden;
    max-height: ${lineHeight * 3}rem;
    white-space: normal;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 3;
  `}

  ${theme.breakpoints.mobile(css`
    padding: ${cardPadding}rem ${cardPadding * 2}rem;
    height: 15.2rem;
    overflow: auto;
  `)}
`;
