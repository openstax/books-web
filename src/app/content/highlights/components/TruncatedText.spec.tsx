import React from 'react';
import renderer from 'react-test-renderer';
import MessageProvider from '../../../MessageProvider';
import TruncatedText from './TruncatedText';

describe('TruncatedText', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<MessageProvider onError={() => null}>
      <TruncatedText text='asdf' isFocused={false} />
    </MessageProvider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when focused', () => {
    const component = renderer.create(<MessageProvider onError={() => null}>
      <TruncatedText text='asdf' isFocused={true} />
    </MessageProvider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows link', async() => {
    const createNodeMock = () => ({
      offsetHeight: 50,
      scrollHeight: 100,
    });
    const component = renderer.create(<MessageProvider onError={() => null}>
      <TruncatedText text='asdf' isFocused={true} />
    </MessageProvider>, {createNodeMock});

    component.update(<MessageProvider onError={() => null}>
      <TruncatedText text='asdf' isFocused={true} />
    </MessageProvider>);

    expect(() => component.root.findByType('span')).not.toThrow();
  });
});
