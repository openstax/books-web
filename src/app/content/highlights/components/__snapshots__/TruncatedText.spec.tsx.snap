// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TruncatedText matches snapshot 1`] = `
<p
  className="TruncatedText-sc-7sqw62-1 fqcHRh"
>
  asdf
</p>
`;

exports[`TruncatedText matches snapshot when focused 1`] = `
<p
  className="TruncatedText-sc-7sqw62-1 epSSBR"
>
  asdf
</p>
`;
